packages/authentication/test/unit/provider.test.ts

```diff
@@ -0,0 +1,88 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {Context, Provider, ValueOrPromise, instantiateClass} from '@loopback/context';
+import {HttpErrors, ParsedRequest} from '@loopback/core';
+import {AuthenticationProvider, getAuthenticatedUser, AuthenticateFn, UserProfile} from '../..';
+import {MockStrategy} from './fixtures/mock-strategy';
+
+describe('AuthenticationProvider', () => {
+  let provider: Provider<AuthenticateFn>;
+  let strategy: MockStrategy;
+
+  const mockUser: UserProfile = {name: 'mock-user', locale: 'mock-locale'};
+
+  beforeEach(givenAuthenticationProvider);
+
+  describe('constructor()', () => {
+    it('instantiateClass injects authentication.strategy in the constructor', async () => {
+      const context: Context = new Context();
+      context.bind('authentication.strategy').to(strategy);
+      provider = await instantiateClass<Provider<AuthenticateFn>>(AuthenticationProvider, context);
+      const authenticationProvider: AuthenticationProvider = provider as AuthenticationProvider;
+      expect(authenticationProvider.getStrategy()).to.be.equal(strategy);
+    });
+  });
+
+  describe('value()', () => {
+    it('returns a function which authenticates a request and returns a user', async () => {
+      const authenticate: AuthenticateFn = await Promise.resolve(provider.value());
+      const request = <ParsedRequest> {};
+      const user = await authenticate(request);
+      expect(user).to.be.equal(mockUser);
+    });
+
+    describe('context.get(provider_key)', () => {
+      it('returns a function which authenticates a request and returns a user', async () => {
+        const context: Context = new Context();
+        context.bind('authentication.strategy').to(strategy);
+        context.bind('authentication.provider').toProvider(AuthenticationProvider);
+        const authenticate = await context.get('authentication.provider');
+        const request = <ParsedRequest> {};
+        const user = await authenticate(request);
+        expect(user).to.be.equal(mockUser);
+      });
+
+      it('throws an error if the injected passport strategy is not valid', async () => {
+        const context: Context = new Context();
+        context.bind('authentication.strategy').to({});
+        context.bind('authentication.provider').toProvider(AuthenticationProvider);
+        const authenticate = await context.get('authentication.provider');
+        const request = <ParsedRequest> {};
+        let error;
+        try {
+          await authenticate(request);
+        } catch (exception) {
+          error = exception;
+        }
+        expect(error.message).to.equal('invalid strategy parameter');
```

Comment:
This will throw `TypeError: Cannot read property 'message' of undefined` in case the code inside `try` block did not throw.

Typically, `should.throw` is used for asserting that a function throws an error, however it seems `should.throw` does not support async functions, thus we need a different solution. My recommendation is to change the initial value of the `error` object:

```ts
let error = { message: 'authenticate() should have failed' };
```